core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java

```diff
@@ -377,154 +372,173 @@ public boolean isClosed() {
+                pingCollection.addPing(pingResponse);
+            }
         }
 
         @Override
         public void close() {
-            if (closed.compareAndSet(false, true)) {
-                receivedResponses.remove(id);
+            List<Connection> toClose = null;
+            synchronized (this) {
+                if (closed.compareAndSet(false, true)) {
+                    activePingingRounds.remove(id);
+                    toClose = new ArrayList<>(tempConnections.values());
+                }
             }
+            if (toClose != null) {
+                // we actually closed
+                try {
+                    pingListener.accept(pingCollection);
+                } finally {
+                    IOUtils.closeWhileHandlingException(toClose);
+                }
+            }
+        }
+
+        public ConnectionProfile getConnectionProfile() {
+            return connectionProfile;
         }
     }
 
 
-    void sendPings(
-        final TimeValue timeout,
-        @Nullable TimeValue waitTime,
-        final SendPingsHandler sendPingsHandler,
-        final List<DiscoveryNode> resolvedDiscoveryNodes) {
+    protected void sendPings(final TimeValue timeout, final PingingRound pingingRound) {
         final UnicastPingRequest pingRequest = new UnicastPingRequest();
-        pingRequest.id = sendPingsHandler.id();
+        pingRequest.id = pingingRound.id();
         pingRequest.timeout = timeout;
         DiscoveryNodes discoNodes = contextProvider.nodes();
 
         pingRequest.pingResponse = createPingResponse(discoNodes);
 
-        HashSet<DiscoveryNode> nodesToPingSet = new HashSet<>();
-        for (PingResponse temporalResponse : temporalResponses) {
-            // Only send pings to nodes that have the same cluster name.
-            if (clusterName.equals(temporalResponse.clusterName())) {
-                nodesToPingSet.add(temporalResponse.node());
-            }
-        }
-        nodesToPingSet.addAll(hostsProvider.buildDynamicNodes());
+        Set<DiscoveryNode> nodesFromResponses = temporalResponses.stream().map(pingResponse -> {
+            assert clusterName.equals(pingResponse.clusterName()) :
+                "got a ping request from a different cluster. expected " + clusterName + " got " + pingResponse.clusterName();
+            return pingResponse.node();
+        }).collect(Collectors.toSet());
 
-        // add all possible master nodes that were active in the last known cluster configuration
-        for (ObjectCursor<DiscoveryNode> masterNode : discoNodes.getMasterNodes().values()) {
-            nodesToPingSet.add(masterNode.value);
-        }
+        // dedup by address
+        final Map<TransportAddress, DiscoveryNode> uniqueNodesByAddress =
+            Stream.concat(pingingRound.getSeedNodes().stream(), nodesFromResponses.stream())
+                .collect(Collectors.toMap(DiscoveryNode::getAddress, n -> n, (n1, n2) -> n1));
```

Comment:
the compiler can't figure out the implied types with the identity. I changed it to be more verbose so it won't look like a typo